### PR TITLE
ARROW-1261: [Java] Add MapVector with reader and writer

### PR DESCRIPTION
**Pull request by BryanCutler**
This adds `MapVector` as a subclass of `ListVector` where the data vector is a Struct with 2 fields: "key" and "value".  A new writer `UnionMapWriter` is added that extends `UnionListWriter` to simplify writing key, value fields. Similarly, the `UnionMapReader` is added to read key, value fields.

### CODE REVIEW

**Comment by BryanCutler** [comment]
This is a WIP, currently need to add roundtrip tests in Java and complete unit tests - and add Javadocs of course :)

**Comment by emkornfield** [comment]
@BryanCutler given its a WIP, do you still want a review?

**Comment by BryanCutler** [comment]
>@BryanCutler given its a WIP, do you still want a review?

@emkornfield it might be better to hold off on a detailed review until I finish up with tests and make another pass through. If you'd like to take a high-level look now and discuss the new classes and APIs, that would be much appreciated! I basically extended everything from a List of key/value structs, but there are other ways to do it too.

**Comment by BryanCutler** [comment]
Making a MapWriter that is able to constrain the StructVector and key Vector to be non-nullable got a little messy.  The vector writers are all designed to only make nullable vectors, so I added a flag that can be set, so that when the writer creates a vector, it can be made as non-nullable. I tried a couple other ways to go about it, like using `NonNullableStructVector`, but it didn't work out very well so this was the cleanest way I could find to go about it. I didn't try to make a MapWriter from scratch, because that seemed overkill, but it could be possible also.

**Comment by BryanCutler** [comment]
@bkietz I think this is ready to try integration tests with. How are things on the C++ side?

**Comment by BryanCutler** [comment]
I think this is ready to review, although still need to work on integration testing. cc @emkornfield @siddharthteotia @pravindra @jacques-n

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: grammar?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: lists?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: Predconditions.checkArgument?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: Preconditions?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java]: not sure this add much without a constant variable?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java
Comment (emkornfield) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java]: Preconditions?

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java
Comment (emkornfield) [inline java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: we should make sure there is a unit or integration test that calls the columns by different names and we can still read/write the type.

File: java/vector/src/main/codegen/templates/UnionMapWriter.java
Comment (emkornfield) [inline java/vector/src/main/codegen/templates/UnionMapWriter.java]: some more javadoc here might be useful.

File: java/vector/src/main/codegen/templates/UnionMapWriter.java
Comment (emkornfield) [inline java/vector/src/main/codegen/templates/UnionMapWriter.java]: should the writer enforce uniqueness/sorted-ness (I suppose this would be difficult in the general case)?

**Review by emkornfield — COMMENTED** [review]
Seems reasonable to me, mostly style comments.  I still haven't fully wrapped my head around readers/writers concept yet so I probably shouldn't be the one to do a final review (will have a better sense of them after I finish my work on Unions in java though, in case no one else has bandwidth to review).

**Comment by bkietz** [comment]
@BryanCutler should be merged soon

**Comment by BryanCutler** [comment]
>@BryanCutler C++ side should be merged soon

Great, thanks @bkietz !  It looks like you have integration tests all ready in your PR, so after you merge then I should be able to enable and run map tests from here?

File: java/vector/src/main/codegen/templates/AbstractFieldWriter.java
Comment (BryanCutler) [reply java/vector/src/main/codegen/templates/AbstractFieldWriter.java]: yes, definitely

File: java/vector/src/main/codegen/templates/StructWriters.java
Comment (BryanCutler) [reply java/vector/src/main/codegen/templates/StructWriters.java]: they are for `DictionaryEncoding` and metadata. If you want to do dictionary encoding, I don't think it is supported with the writers. If `metadata` is initialized to null, it creates an empty map.

File: java/vector/src/main/codegen/templates/UnionMapWriter.java
Comment (BryanCutler) [reply java/vector/src/main/codegen/templates/UnionMapWriter.java]: I think that was discussed somewhere else, and was decided it is up to the application to ensure these things. The `sortedKeys` field is just used as a hint

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: not sure what you are referring to?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: sure, that would be cleaner

File: java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java
Comment (BryanCutler) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/NonNullableStructVector.java]: The changes in this class were moved from the subclass, for when I was trying to use this class as a FieldVector for the map data. It didn't work out, so I don't think I can revert these but it is actually the same in all vectors, which I agree could be cleaned up.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java
Comment (BryanCutler) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: Yeah, that would be good. I'll add it.

**Comment by BryanCutler** [comment]
Thanks for reviewing @emkornfield , I made some updates but still need to add the test you suggested. I don't normally use the reader/writer classes so it would be good to have maybe @pravindra or @siddharthteotia take a look at these.

**Comment by tvamsikalyan** [comment]
Hi Bryan, Greetings!
@pravindra asked to me to review this PR once.
I will review this PR over this weekend. If you need it before weekend, please let me know.
Thank you so much.
T. Vamsi Kalyan.

File: java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java
Comment (BryanCutler) [reply java/vector/src/test/java/org/apache/arrow/vector/ipc/TestArrowFile.java]: done

**Review by tvamsikalyan — APPROVED** [review]
looks good to me.
Only comment I have is one test case with nested types, example: map<int, list<string>> would be useful in my opinion

**Comment by BryanCutler** [comment]
Thanks @tvamsikalyan for the reveiew!  I do have a test for reading/writing map<long, list> here https://github.com/apache/arrow/pull/4444/files#diff-4ca74488fffe65225bb3faf300664b23R482

**Comment by tvamsikalyan** [comment]
> Thanks @tvamsikalyan for the reveiew! I do have a test for reading/writing map<long, list> here https://github.com/apache/arrow/pull/4444/files#diff-4ca74488fffe65225bb3faf300664b23R482
I see it now. Thank you so much for the reply and for the link.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java
Comment (emkornfield) [reply java/vector/src/main/java/org/apache/arrow/vector/complex/MapVector.java]: I think I wanted entries->lists, but I think this is probably ok.

**Comment by emkornfield** [comment]
+1 LGTM

**Pull request closed without merging**
